metron-analytics/metron-profiler-common/src/main/java/org/apache/metron/profiler/hbase/DecodableRowKeyBuilder.java

```diff
@@ -0,0 +1,402 @@
+        groups.add(group);
+      }
+
+      // decode the period
+      long periodId = buffer.getLong();
+      long duration = buffer.getLong();
+      ProfilePeriod period = ProfilePeriod.buildFromPeriod(periodId, duration, TimeUnit.MILLISECONDS);
+
+      return new ProfileMeasurement()
+              .withProfileName(profile)
+              .withEntity(entity)
+              .withGroups(groups)
+              .withProfilePeriod(period);
+
+    } catch(BufferUnderflowException e) {
+      throw new IllegalArgumentException("Unable to decode the row key", e);
+    }
+
+  }
+
+  /**
+   * Builds the 'group' component of the row key.
+   * @param groups The groups to include in the row key.
+   */
+  private byte[] encodeGroups(List<Object> groups) {
+
+    // encode each of the groups and determine their size
+    int lengthOfGroups = 0;
+    List<byte[]> groupBytes = new ArrayList<>();
+    for(Object group : groups) {
+      byte[] groupB = Bytes.toBytes(String.valueOf(group));
+      groupBytes.add(groupB);
+      lengthOfGroups += groupB.length;
+    }
+
+    // encode each of the groups
+    ByteBuffer buffer = ByteBuffer
+            .allocate((Integer.BYTES * (1 + groups.size())) + lengthOfGroups)
+            .order(byteOrder)
+            .putInt(groups.size());
+
+    for(byte[] groupB : groupBytes) {
+      buffer.putInt(groupB.length).put(groupB);
+    }
+
+    return buffer.array();
+  }
+
+  /**
+   * Builds the 'time' portion of the row key
+   * @param period The ProfilePeriod in which the ProfileMeasurement was taken.
+   */
+  private static byte[] encodePeriod(ProfilePeriod period) {
+    return encodePeriod(period.getPeriod());
+  }
+
+  /**
+   * Builds the 'time' portion of the row key
+   * @param period the period
+   */
+  private static byte[] encodePeriod(long period) {
+    return Bytes.toBytes(period);
+  }
+
+  /**
+   * Calculates a salt value that is used as part of the row key.
+   *
+   * The salt is calculated as 'md5(period) % N' where N is a configurable value that ideally
+   * is close to the number of nodes in the Hbase cluster.
+   *
+   * @param period The period in which a profile measurement is taken.
+   */
+  public static byte[] encodeSalt(ProfilePeriod period, int saltDivisor) {
+    return encodeSalt(period.getPeriod(), saltDivisor);
+  }
+
+  /**
+   * Calculates a salt value that is used as part of the row key.
+   *
+   * The salt is calculated as 'md5(period) % N' where N is a configurable value that ideally
+   * is close to the number of nodes in the Hbase cluster.
+   *
+   * @param period The period
+   * @param saltDivisor The salt divisor
+   */
+  public static byte[] encodeSalt(long period, int saltDivisor) {
+    try {
+      // an MD5 is 16 bytes aka 128 bits
+      MessageDigest digest = MessageDigest.getInstance("MD5");
+      byte[] hash = digest.digest(encodePeriod(period));
+      int salt = Bytes.toShort(hash) % saltDivisor;
+      return Bytes.toBytes(salt);
```

Comment:
\@ All, please note that my overall review comment (as opposed to these detailed comments) weren't published to email by github.  Please see it at https://github.com/apache/metron/pull/622#pullrequestreview-51932717 
It does address the issue of changed profile specs being different profiles.

@simonellistonball , your question sort of brings up the issue of why we use salts at all.  Apparently the leading bytes are important; if they weren't there would be no reason to use a salt, but we could just trust in the later bytes to sufficiently differentiate the partition/region for the row.  As I understand hbase table partitioning (imperfectly), the more times partitioning has happened, and the broader, more well-mixed set of rowkeys are in the table, then the more likely that rowkeys will end up in the same partition (region) unless they differ early in the byte array.  So, if the hbase instance mostly only has data with similar rowkeys, then the bytes later in the rowkey, such as the profile name, will serve sufficiently.  But as the table fills up with a broader range of rowkeys, as it will with our salting strategy, and as it grows larger over time, then a bunch of inserts with the same salt (leading few bytes) is more likely to land in the same partition.

I'm not an hbase expert, so if you feel I've got this wrong, please clarify.